File: 02-observer/weather-node/displays/current.ts
import Subject from "../weatherdata/subject";
import DisplayElement from "./display";
import Observer from "./observer";

export default class CurrentConditionsDisplay
  implements Observer, DisplayElement
{
  private temperature: number;
  private humidity: number;
  private weatherData: Subject;

  constructor(weatherData: Subject) {
    this.weatherData = weatherData;
    this.temperature = 0;
    this.humidity = 0;
    this.weatherData.registerObservers(this);
  }

  public update(temp: number, humidity: number, pressure: number): void {
    this.humidity = humidity;
    this.temperature = temp;
    this.display();
  }

  public display(): void {
    console.log(
      `Current conditions: ${this.temperature}F degrees and ${this.humidity}% humidity.`
    );
  }
}

File: 02-observer/weather-go/display/current.go
package display

import (
	"fmt"

	"github.com/k-nox/design-patterns/02-observer/weather-go/observer"
)

type CurrentConditions struct {
	temperature float32
	humidity float32
	weatherData observer.Subject
}

func (c *CurrentConditions) Update(temp float32, humidity float32, pressure float32) {
	c.temperature = temp
	c.humidity = humidity
	c.Display()
}

func (c *CurrentConditions) Display() {
	fmt.Printf("Current conditions: %.2fF degrees and %.2f%% humidity\n", c.temperature, c.humidity)
}

func NewCurrentConditions(wd observer.Subject) *CurrentConditions {
	cc := &CurrentConditions{
		weatherData: wd,
	}

	wd.RegisterObserver(cc)
	return cc
}
